src/main/java/com/bookingTour/controller/TourDetailsController.java

```diff
@@ -30,10 +32,25 @@
     @Autowired
     private TourService tourService;
 
-    @GetMapping(value = { "/add" })
-    public String add(@PathVariable(name = "tour_id", required = true)  @Min(1) Long tourId, Locale locale, Model model) {
+    @GetMapping(value = {"/tours/{tour_id}/tour-details"})
+    public String index(@RequestParam(name = "page", required = false) Optional<Integer> page,
+                        @PathVariable(name = "tour_id", required = true) Long tourId,
+                        Locale locale, Model model, HttpServletRequest request) {
+        TourDetailInfo tourDetailInfo = new TourDetailInfo();
+        tourDetailInfo.setPage(page.orElse(1));
+        tourDetailInfo.setTourId(tourId);
+        Page<TourDetailInfo> tourDetails = tourDetailService.paginate(tourDetailInfo);
+        model.addAttribute("tourDetails", tourDetails);
+        tourDetailInfo.setTourId(tourId);
+        model.addAttribute("tourInfo", tourService.findTour(tourId));
+        model.addAttribute("tourDetailSearch", tourDetailInfo);
+        return "tour-details/index";
+    }
+
+    @GetMapping(value = {"/admin/tours/{tour_id}/tour-details/add"})
```

Comment:
role admin mà chị chưa thấy việc check user đã login hay chưa nè

và xử lý check login hay chưa sẽ dùng rát nhìu nơi nên em nên viết nó ở class Base Controller để có thể tái sử dụng nè

Reply:
ngoài ra, chị ko khuyến khích xử lý cú role admin và user thường vào 1 controller mô, trên thwucj tế sẽ tách biệt 2 controller á, 
thường nhé
1 table - 1 model - 1 Dao - 1 service - x controller 
nếu project có 2 quyền thì X chính là 2 ...

Reply:
dạ em có thêm config cho security bắt buộc các request tới phải có quyền tức là đã login ạ, nếu chưa có quyền thì sẽ trả về trang login ạ  
`.anyRequest().hasAnyAuthority(Role.USER.name(), Role.ADMIN.name())`

Reply:
> role admin mà chị chưa thấy việc check user đã login hay chưa nè
> 
> và xử lý check login hay chưa sẽ dùng rát nhìu nơi nên em nên viết nó ở class Base Controller để có thể tái sử dụng nè

dạ em có thêm config cho security bắt buộc các request tới phải có quyền tức là đã login ạ, nếu chưa có quyền thì sẽ trả về trang login ạ
.anyRequest().hasAnyAuthority(Role.USER.name(), Role.ADMIN.name())

Về check role admin thì em có accessDeniedHandler để trả về trang accessDenied khi user truy cập tới tất cả các trang bắt đầu bằng admin ạ

`.and().exceptionHandling().accessDeniedHandler(accessDeniedHandler)`

![image](https://user-images.githubusercontent.com/47592603/92423130-1f4b8a00-f1aa-11ea-9dd5-6b133df336b1.png)